### PR TITLE
[TEST] Added unit tests where possible

### PR DESCRIPTION
**Pull request by mattyb149**
Added defensive code where it was low-risk and allowed the unit tests to pass successfully.  Also added Mockito as a test dependency.

Lots of UI here, could probably eke out more coverage if necessary

### CODE REVIEW

**Comment by lucboudreau** [comment]
Build Commands:
==============
Build:
------
1. rm -r ~/.kettle; rm -r ~/.pentaho; echo removed hidden folders
2. ant -Djunit.jvmargs=-XX:PermSize=256m -f dbdialog/build.xml clean-all resolve jacoco jacoco-integration checkstyle publish-local

Cleanup:
-------
rm -r ~/.ivy2/local || echo no publish local to remove

Changed files:
==============
  dbdialog/ivy.xml
  dbdialog/src/org/pentaho/ui/database/DatabaseConnectionDialog.java
  dbdialog/src/org/pentaho/ui/database/event/DataHandler.java
  dbdialog/src/org/pentaho/ui/database/event/FragmentHandler.java
  dbdialog/test-src/org/pentaho/ui/database/DatabaseConnectionDialogTest.java
  dbdialog/test-src/org/pentaho/ui/database/MessagesTest.java
  dbdialog/test-src/org/pentaho/ui/database/event/DataHandlerTest.java
  dbdialog/test-src/org/pentaho/ui/database/event/FragmentHandlerTest.java

Checkstyle violations:
======================
File dbdialog/src/org/pentaho/ui/database/event/DataHandler.java
1. Message: 'abstract' modifier out of order with the JLS suggestions. at Line: 1459 Column: 20

File dbdialog/test-src/org/pentaho/ui/database/event/DataHandlerTest.java
1. Message: Unused import - org.mockito.Matchers.anyBoolean. at Line: 26 Column: 15
2. Message: Unused import - org.pentaho.di.core.exception.KettleDatabaseException. at Line: 8 Column: 8

File dbdialog/test-src/org/pentaho/ui/database/event/FragmentHandlerTest.java
1. Message: Unused import - org.mockito.Mockito.doReturn. at Line: 18 Column: 15
2. Message: Unused import - org.mockito.Mockito.spy. at Line: 20 Column: 15
3. Message: Unused import - org.pentaho.ui.xul.XulLoader. at Line: 9 Column: 8


License header violations:
==========================
1. dbdialog/test-src/org/pentaho/ui/database/event/DataHandlerTest.java
2. dbdialog/test-src/org/pentaho/ui/database/event/FragmentHandlerTest.java







Unit test coverage change:
============
org.pentaho.ui.database.event.DataHandler:
 * branchPercentCoverage :28.11%
 * linePercentCoverage :46.73%
 * methodPercentCoverage :72.97%
 * complexityPercentCoverage :15.75%
 * instructionPercentCoverage :42.07%
    
org.pentaho.ui.database.event.DataHandler.DatabaseTypeListener:
 * branchPercentCoverage :50.00%
 * linePercentCoverage :95.83%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :80.00%
 * instructionPercentCoverage :96.70%
    
org.pentaho.ui.database.Messages:
 * branchPercentCoverage :50.00%
 * linePercentCoverage :88.00%
 * methodPercentCoverage :87.50%
 * complexityPercentCoverage :77.78%
 * instructionPercentCoverage :96.00%
    
org.pentaho.ui.util.Launch.Status:
 * linePercentCoverage :100.00%
 * methodPercentCoverage :50.00%
 * complexityPercentCoverage :50.00%
 * instructionPercentCoverage :76.32%
    
org.pentaho.ui.database.event.DataHandler.new DataHandler.DatabaseTypeListener() {...}:
 * linePercentCoverage :11.11%
 * methodPercentCoverage :25.00%
 * complexityPercentCoverage :16.67%
 * instructionPercentCoverage :30.30%
    
org.pentaho.ui.database.event.FragmentHandler:
 * branchPercentCoverage :77.27%
 * linePercentCoverage :93.06%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :75.00%
 * instructionPercentCoverage :90.84%
    
org.pentaho.ui.util.Launch:
 * branchPercentCoverage :58.33%
 * linePercentCoverage :59.09%
 * methodPercentCoverage :50.00%
 * complexityPercentCoverage :25.00%
 * instructionPercentCoverage :58.70%
    
org.pentaho.ui.database.DatabaseConnectionDialog:
 * linePercentCoverage :33.33%
 * methodPercentCoverage :66.67%
 * complexityPercentCoverage :50.00%
 * instructionPercentCoverage :26.79%
    
org.pentaho.ui.database.event.DataHandler.new PropertyChangeListener() {...}:
 * linePercentCoverage :25.00%
 * methodPercentCoverage :50.00%
 * complexityPercentCoverage :33.33%
 * instructionPercentCoverage :40.00%

**Comment by lucboudreau** [comment]
Build Commands:
==============
Build:
------
1. rm -r ~/.kettle; rm -r ~/.pentaho; echo removed hidden folders
2. ant -Djunit.jvmargs=-XX:PermSize=256m -f dbdialog/build.xml clean-all resolve jacoco jacoco-integration checkstyle publish-local

Cleanup:
-------
rm -r ~/.ivy2/local || echo no publish local to remove

Changed files:
==============
  dbdialog/ivy.xml
  dbdialog/src/org/pentaho/ui/database/DatabaseConnectionDialog.java
  dbdialog/src/org/pentaho/ui/database/event/DataHandler.java
  dbdialog/src/org/pentaho/ui/database/event/FragmentHandler.java
  dbdialog/test-src/org/pentaho/ui/database/DatabaseConnectionDialogTest.java
  dbdialog/test-src/org/pentaho/ui/database/MessagesTest.java
  dbdialog/test-src/org/pentaho/ui/database/event/DataHandlerTest.java
  dbdialog/test-src/org/pentaho/ui/database/event/FragmentHandlerTest.java

Checkstyle violations:
======================
File dbdialog/src/org/pentaho/ui/database/event/DataHandler.java
1. Message: 'abstract' modifier out of order with the JLS suggestions. at Line: 1459 Column: 20

File dbdialog/test-src/org/pentaho/ui/database/event/DataHandlerTest.java
1. Message: Unused import - org.mockito.Matchers.anyBoolean. at Line: 26 Column: 15
2. Message: Unused import - org.pentaho.di.core.exception.KettleDatabaseException. at Line: 8 Column: 8

File dbdialog/test-src/org/pentaho/ui/database/event/FragmentHandlerTest.java
1. Message: Unused import - org.mockito.Mockito.doReturn. at Line: 18 Column: 15
2. Message: Unused import - org.mockito.Mockito.spy. at Line: 20 Column: 15
3. Message: Unused import - org.pentaho.ui.xul.XulLoader. at Line: 9 Column: 8


License header violations:
==========================
1. dbdialog/test-src/org/pentaho/ui/database/event/DataHandlerTest.java
2. dbdialog/test-src/org/pentaho/ui/database/event/FragmentHandlerTest.java







Unit test coverage change:
============
org.pentaho.ui.database.event.DataHandler:
 * branchPercentCoverage :28.11%
 * linePercentCoverage :46.73%
 * methodPercentCoverage :72.97%
 * complexityPercentCoverage :15.75%
 * instructionPercentCoverage :42.07%
    
org.pentaho.ui.database.event.DataHandler.DatabaseTypeListener:
 * branchPercentCoverage :50.00%
 * linePercentCoverage :95.83%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :80.00%
 * instructionPercentCoverage :96.70%
    
org.pentaho.ui.database.Messages:
 * branchPercentCoverage :50.00%
 * linePercentCoverage :88.00%
 * methodPercentCoverage :87.50%
 * complexityPercentCoverage :77.78%
 * instructionPercentCoverage :96.00%
    
org.pentaho.ui.util.Launch.Status:
 * linePercentCoverage :100.00%
 * methodPercentCoverage :50.00%
 * complexityPercentCoverage :50.00%
 * instructionPercentCoverage :76.32%
    
org.pentaho.ui.database.event.DataHandler.new DataHandler.DatabaseTypeListener() {...}:
 * linePercentCoverage :11.11%
 * methodPercentCoverage :25.00%
 * complexityPercentCoverage :16.67%
 * instructionPercentCoverage :30.30%
    
org.pentaho.ui.database.event.FragmentHandler:
 * branchPercentCoverage :77.27%
 * linePercentCoverage :93.06%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :75.00%
 * instructionPercentCoverage :90.84%
    
org.pentaho.ui.util.Launch:
 * branchPercentCoverage :58.33%
 * linePercentCoverage :59.09%
 * methodPercentCoverage :50.00%
 * complexityPercentCoverage :25.00%
 * instructionPercentCoverage :58.70%
    
org.pentaho.ui.database.DatabaseConnectionDialog:
 * linePercentCoverage :33.33%
 * methodPercentCoverage :66.67%
 * complexityPercentCoverage :50.00%
 * instructionPercentCoverage :26.79%
    
org.pentaho.ui.database.event.DataHandler.new PropertyChangeListener() {...}:
 * linePercentCoverage :25.00%
 * methodPercentCoverage :50.00%
 * complexityPercentCoverage :33.33%
 * instructionPercentCoverage :40.00%

File: dbdialog/src/org/pentaho/ui/database/DatabaseConnectionDialog.java [hunk 1]
```diff
@@ -34,7 +34,7 @@
 
   public static final String DIALOG_DEFINITION_FILE = "org/pentaho/ui/database/databasedialog.xul";
 
-  private Map<String, String> extendedClasses = new HashMap<String, String>();
+  protected Map<String, String> extendedClasses = new HashMap<>();
```
Comment (lucboudreau) [inline dbdialog/src/org/pentaho/ui/database/DatabaseConnectionDialog.java]: I think we're not getting rid of these just yet.

File: dbdialog/src/org/pentaho/ui/database/event/DataHandler.java [hunk 2]
```diff
@@ -782,18 +786,20 @@ private void setInfo( DatabaseMeta meta ) {
...
+    if ( connectionNameBox != null ) {
+      connectionNameBox.setValue( meta.getDisplayName() );
+    }
 
     PluginRegistry registry = PluginRegistry.getInstance();
     PluginInterface dInterface = registry.getPlugin( DatabasePluginType.class, meta.getPluginId() );
 
     // Connection type:
-    int index = new ArrayList<String>( connectionMap.keySet() ).indexOf( dInterface.getName() );
+    int index = ( dInterface == null ? -1 : new ArrayList<>( connectionMap.keySet() ).indexOf( dInterface.getName() ) );
```
Comment (lucboudreau) [inline dbdialog/src/org/pentaho/ui/database/event/DataHandler.java]: Same comment about generics here.

File: dbdialog/src/org/pentaho/ui/database/DatabaseConnectionDialog.java (hunk 1 above)
Comment (mattyb149) [inline dbdialog/src/org/pentaho/ui/database/DatabaseConnectionDialog.java]: The diamond operator is a Java 7 feature, @dkincade said we can use Java 7 features in master now

File: dbdialog/test-src/org/pentaho/ui/database/event/FragmentHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    when( component.getParent() ).thenReturn( parent );
+    when( document.getElementById( "database-options-box" ) ).thenReturn( component );
+    when( xulDomContainer.loadFragment( anyString(), any() ) ).thenThrow( new XulException() );
+    fragmentHandler.loadDatabaseOptionsFragment( null );
+  }
+
+  @Test
+  public void testGetFragment() throws Exception {
+    DatabaseInterface dbInterface = mock( DatabaseInterface.class );
+    assertEquals( "org/pentaho/ui/database/", fragmentHandler.getFragment( dbInterface, null, null, null ) );
```
Comment (lucboudreau) [inline dbdialog/test-src/org/pentaho/ui/database/event/FragmentHandlerTest.java]: Can we test the other parameters of getFragment?

Comment (mattyb149) [inline dbdialog/test-src/org/pentaho/ui/database/event/FragmentHandlerTest.java]: We could, I just stopped when I hit 100% coverage

Comment (lucboudreau) [inline dbdialog/test-src/org/pentaho/ui/database/event/FragmentHandlerTest.java]: OK then. We probably hit non-null scenarios elsewhere.

**Comment by mattyb149** [comment]
@lucboudreau Updated to incorporate comments (including removing Checkstyle violations, etc.)

**Comment by lucboudreau** [comment]
Build Commands:
==============
Build:
------
1. rm -r ~/.kettle; rm -r ~/.pentaho; echo removed hidden folders
2. ant -Djunit.jvmargs=-XX:PermSize=256m -f dbdialog/build.xml clean-all resolve jacoco jacoco-integration checkstyle publish-local

Cleanup:
-------
rm -r ~/.ivy2/local || echo no publish local to remove

Changed files:
==============
  dbdialog/ivy.xml
  dbdialog/src/org/pentaho/ui/database/DatabaseConnectionDialog.java
  dbdialog/src/org/pentaho/ui/database/event/DataHandler.java
  dbdialog/src/org/pentaho/ui/database/event/FragmentHandler.java
  dbdialog/test-src/org/pentaho/ui/database/DatabaseConnectionDialogTest.java
  dbdialog/test-src/org/pentaho/ui/database/MessagesTest.java
  dbdialog/test-src/org/pentaho/ui/database/event/DataHandlerTest.java
  dbdialog/test-src/org/pentaho/ui/database/event/FragmentHandlerTest.java

Checkstyle violations:
======================
No errors found

License header violations:
==========================
No violations found








Unit test coverage change:
============
org.pentaho.ui.database.event.DataHandler:
 * branchPercentCoverage :28.11%
 * linePercentCoverage :46.73%
 * methodPercentCoverage :72.97%
 * complexityPercentCoverage :15.75%
 * instructionPercentCoverage :42.07%
    
org.pentaho.ui.database.event.DataHandler.DatabaseTypeListener:
 * branchPercentCoverage :50.00%
 * linePercentCoverage :95.83%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :80.00%
 * instructionPercentCoverage :96.70%
    
org.pentaho.ui.database.Messages:
 * branchPercentCoverage :50.00%
 * linePercentCoverage :88.00%
 * methodPercentCoverage :87.50%
 * complexityPercentCoverage :77.78%
 * instructionPercentCoverage :96.00%
    
org.pentaho.ui.util.Launch.Status:
 * linePercentCoverage :100.00%
 * methodPercentCoverage :50.00%
 * complexityPercentCoverage :50.00%
 * instructionPercentCoverage :76.32%
    
org.pentaho.ui.database.event.DataHandler.new DataHandler.DatabaseTypeListener() {...}:
 * linePercentCoverage :11.11%
 * methodPercentCoverage :25.00%
 * complexityPercentCoverage :16.67%
 * instructionPercentCoverage :30.30%
    
org.pentaho.ui.database.event.FragmentHandler:
 * branchPercentCoverage :77.27%
 * linePercentCoverage :93.06%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :75.00%
 * instructionPercentCoverage :90.84%
    
org.pentaho.ui.util.Launch:
 * branchPercentCoverage :58.33%
 * linePercentCoverage :59.09%
 * methodPercentCoverage :50.00%
 * complexityPercentCoverage :25.00%
 * instructionPercentCoverage :58.70%
    
org.pentaho.ui.database.DatabaseConnectionDialog:
 * linePercentCoverage :33.33%
 * methodPercentCoverage :66.67%
 * complexityPercentCoverage :50.00%
 * instructionPercentCoverage :26.79%
    
org.pentaho.ui.database.event.DataHandler.new PropertyChangeListener() {...}:
 * linePercentCoverage :25.00%
 * methodPercentCoverage :50.00%
 * complexityPercentCoverage :33.33%
 * instructionPercentCoverage :40.00%

**Pull request merged**